types/multer/index.d.ts

```diff
@@ -1,127 +1,305 @@
-// Type definitions for multer 1.3
+// Type definitions for multer 1.4
 // Project: https://github.com/expressjs/multer
 // Definitions by: jt000 <https://github.com/jt000>
 //                 vilicvane <https://vilic.github.io/>
 //                 David Broder-Rodgers <https://github.com/DavidBR-SW>
 //                 Michael Ledin <https://github.com/mxl>
 //                 HyunSeob Lee <https://github.com/hyunseob>
 //                 Pierre Tchuente <https://github.com/PierreTchuente>
+//                 Oliver Emery <https://github.com/thrymgjol>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
-// TypeScript Version: 2.3
+// TypeScript Version: 2.8
 
-import * as express from 'express';
+import { Request, RequestHandler } from 'express';
+import { Readable } from 'stream';
+
+declare global {
+    namespace Express {
+        namespace Multer {
+            /** Object containing file metadata and access information. */
+            interface File {
+                /** Name of the form field associated with this file. */
+                fieldname: string;
+                /** Name of the file on the uploader's computer. */
+                originalname: string;
+                /**
+                 * Value of the `Content-Transfer-Encoding` header for this file.
+                 * @deprecated since July 2015
+                 * @see RFC 7578, Section 4.7
+                 */
+                encoding: string;
+                /** Value of the `Content-Type` header for this file. */
+                mimetype: string;
+                /** Size of the file in bytes. */
+                size: number;
+                /** `DiskStorage` Directory to which this file has been uploaded. */
+                destination: string;
+                /** `DiskStorage` Name of this file within `destination`. */
+                filename: string;
+                /** `DiskStorage` Full path to the uploaded file. */
+                path: string;
+                /** `MemoryStorage` A Buffer containing the entire file. */
+                buffer: Buffer;
+                /** A readable stream of this file. */
+                stream: Readable;
+            }
+        }
+
+        interface Request {
+            /** `Multer.File` object populated by `single()` middleware. */
+            file: Multer.File;
+            /**
+             * Array or dictionary of `Multer.File` object populated by `array()`,
+             * `fields()`, and `any()` middleware.
+             */
+            files: {
+                [fieldname: string]: Multer.File[];
+            } | Multer.File[];
+        }
+    }
+}
+
+declare class Multer {
```

Comment:
@	weswigham I think this class has to be `export`ed, or else there is no way to write this code:


```
import multer from 'multer';

class Foo {
  multerFactory: ???;

  constructor() {
    this.multerFactory = multer({...});
  }  
}
```

Prior to this PR, we had `multer.Instance`, but this patch removes that from the exported namespace

Reply:
Re-adding something like `export type Instance = Multer` is something that could be done.